src/score_gen/score.rs

```diff
@@ -1,162 +1,542 @@
+    // Safe to unwrap here.
+    let score = bid.score.unwrap();
+    let r1 = composer.add_input(score.r1);
+    let r2 = composer.add_input(score.r2);
+    let y = composer.add_input(score.y);
+    let y_prime = composer.add_input(score.y_prime);
+    let score_var = composer.add_input(score.score);
+    let two_pow_128 = Scalar::from(2u64).pow(&[128, 0, 0, 0]);
+    let two_pow_128_buint = BigUint::from_bytes_le(&two_pow_128.to_bytes());
+    // 1. Y = 2^128 * r1 + Y'
+    composer.add_gate(
+        y_prime,
+        r1,
+        y,
+        Scalar::one(),
+        two_pow_128,
+        -Scalar::one(),
+        Scalar::zero(),
+        Scalar::zero(),
+    );
+    // 2.(r1 < |Fr|/2^128 AND Y' < 2^128) OR (r1 = |Fr|/2^128 AND Y' < |Fr| mod 2^128).
     //
-    // We don't need the input since it should be 0. If it is not, the verification process
-    // will fail.
-    composer.mul(
+    // 2.1. First op will be a complex rangeproof between r1 and the range (Order of the Scalar Field / 2^128 (No modular division))
+    // The result should be 0 if the rangeproof holds.
+    let scalar_field_ord_div_2_pow_128 = {
+        let scalar_field_order =
+            BigUint::from_bytes_le(&(-Scalar::one()).to_bytes()) + BigUint::one();
+        scalar_field_order / &two_pow_128_buint
+    };
+    let first_cond = single_complex_range_proof(
+        composer,
+        score.r1,
+        biguint_to_scalar(scalar_field_ord_div_2_pow_128.clone())?,
+    )?;
+
+    // 2.2. Then we have a single Rangeproof between Y' being in the range [0-2^128]
+    let second_cond = single_complex_range_proof(composer, score.y_prime, two_pow_128)?;
+    // 2.3. Third, we have an equalty checking between r1 & the order of the Scalar field divided (no modular division)
+    // by 2^128.
+    // We simply subtract both values and if it's equal, we will get a 0.
+
+    let one = composer.add_input(Scalar::one());
+    let third_cond = {
+        let zero_or_other = composer.add(
+            (Scalar::one(), r1),
+            (
+                -biguint_to_scalar(scalar_field_ord_div_2_pow_128.clone())?,
+                one,
+            ),
+            Scalar::zero(),
+            Scalar::zero(),
+        );
+        let u = score.r1 - biguint_to_scalar(scalar_field_ord_div_2_pow_128)?;
+        // Conditionally assign `1` or `0` to `y`.
+        let y = if u == Scalar::zero() {
+            composer.add_input(Scalar::one())
+        } else {
+            composer.add_input(Scalar::zero())
+        };
+        // Conditionally assign `1/u` or `0` to z
+        let mut z = composer.zero_var;
+        if u != Scalar::zero() {
+            // If u != zero -> `z = 1/u`
+            // Otherways, `u = 0` as it was defined avobe.
+            // Check inverse existance, otherways, err.
+            if u.invert().is_none().into() {
+                return Err(ScoreError::NonExistingInverse.into());
+            };
+            // Safe to unwrap here.
```

Comment:
again, this is providing no information, just stating something.

Reply:
Added an explanation.

This is needed because `u` might not have inverse. And the unwrap will cause a failure.